### PR TITLE
Fix a bug where exceptions raised in decorators are not handled by exception handlers of annotated services

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

While working on #3674, there was a regression that an exception raised
in a decorator was not properly handled by `ExceptionHandlerFunction`
attached to the annotated services.

Modifications:

- Add `ExceptionHandlingHttpService` that handles an exception using
  `ExceptionHandlerFunction`s
  The decorator wraps the annotated service after all user decorators are applied.

Result:

`ExceptionHandlerFunction` of annotated services now correctly handles
causes raised by service decorators.

### CODE REVIEW

**Review by trustin — APPROVED** [review]
:+1:

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java [hunk 1]
```diff
@@ -492,6 +493,30 @@ public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req,
...
+        ExceptionHandlingHttpService(HttpService service, ExceptionHandlerFunction exceptionHandler) {
+            super(service);
+            this.exceptionHandler = exceptionHandler;
+        }
+
+        @Override
+        public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) {
+            try {
+                final HttpResponse response = unwrap().serve(ctx, req);
+                return response.recover(cause -> {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java:509]: I'm also wondering if we need to capture this `cause`. Probably not. 😄 
Let's do that if there's a demand for that.

**Review by minwoox — APPROVED** [review]
Thanks!

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java:509]: We can think of exception handlers attached as domain exception handlers that can fully recover domain exceptions to successful responses.
So I think it is fine as it is.

**Pull request merged**